File: HW1/Q1.jl
using JuMP
using Gurobi   

num_engineerss = 3
num_projects  = 4

cap = [
    [90, 80, 10, 50],  # engineers 1
    [60, 70, 50, 65],  # engineers 2
    [70, 40, 80, 85]   # engineers 3
]
required = [70, 50, 85, 35]  # required hours for each project
max_hours = 80

# ============================================================
#                           Model
# ============================================================
model = Model(Gurobi.Optimizer)

# Decision variables x[i,j]
@variable(model, x[1:num_engineerss, 1:num_projects] >= 0)

# Objective: maximize sum of cap[i][j] * x[i,j]
@objective(model, Max, sum(cap[i][j]*x[i,j] for i in 1:num_engineerss, j in 1:num_projects))

# Constraints:
# 1) Each project j requires exactly required[j] total hours
for j in 1:num_projects
    @constraint(model, sum(x[i,j] for i in 1:num_engineerss) == required[j])
end

# 2) Each engineers i can use at most 80 hours
for i in 1:num_engineerss
    @constraint(model, sum(x[i,j] for j in 1:num_projects) <= max_hours)
end

# ============================================================
#                           Solver
# ============================================================
optimize!(model)

# ============================================================
#                           Output
# ============================================================
println("Objective value (Total Capability): ", objective_value(model))
println("Decision variables x[i,j]:")
for i in 1:num_engineerss
    for j in 1:num_projects
        println("Engineer: $i spend on project: $j = ", value(x[i,j]))
    end
end

File: HW1/Q1.py
from pyomo.environ import *

num_engineers = 3
num_projects  = 4

cap = {
    (1, 1): 90, (1, 2): 80, (1, 3): 10, (1, 4): 50,
    (2, 1): 60, (2, 2): 70, (2, 3): 50, (2, 4): 65,
    (3, 1): 70, (3, 2): 40, (3, 3): 80, (3, 4): 85
}

required = {1: 70, 2: 50, 3: 85, 4: 35}
max_hours = 80

# ============================================================
#                           Model
# ============================================================

model = ConcreteModel()

# Sets for engineers and projects
model.engineers = RangeSet(1, num_engineers)
model.projects = RangeSet(1, num_projects)

# Decision variables: x[i,j], hours designer i spends on project j
model.x = Var(model.engineers, model.projects, domain=NonNegativeReals)

# Objective function: maximize total score * hours
model.obj = Objective(
    expr=sum(cap[i, j] * model.x[i, j] for i in model.engineers for j in model.projects),
    sense=maximize
)

# Constraints:
# 1. Each project must meet its required hours
def project_hours_rule(model, j):
    return sum(model.x[i, j] for i in model.engineers) == required[j]
model.project_hours = Constraint(model.projects, rule=project_hours_rule)

# 2. Each designer can work at most 80 hours in total
def engineers_hours_rule(model, i):
    return sum(model.x[i, j] for j in model.projects) <= max_hours
model.engineers_hours = Constraint(model.engineers, rule=engineers_hours_rule)

# ============================================================
#                           Solver
# ============================================================
solver = SolverFactory('gurobi')
result = solver.solve(model, tee=True)

# ============================================================
#                           Output
# ============================================================
print("\nSolver Status:", result.solver.status)
print("Termination Condition:", result.solver.termination_condition)
print("\nObjective (Total Capability):", model.obj())

for i in model.engineers:
    for j in model.projects:
        print(f"Engineer {i} on Project {j}: {model.x[i, j].value:.2f} hours")